mmv1/templates/terraform/constants/integration_connectors_connection.go.erb

```diff
@@ -0,0 +1,21 @@
+<% unless compiler == "terraformgoogleconversion-codegen" -%>
+// waitforConnectionReady waits for an connecion to leave the
+// "CREATING" state, to indicate that it's ready.
+func waitforConnectionReady(d *schema.ResourceData, config *transport_tpg.Config, timeout time.Duration) error {
+	return resource.Retry(timeout, func() *resource.RetryError {
+		if err := resourceIntegrationConnectorsConnectionRead(d, config); err != nil {
+			return resource.NonRetryableError(err)
+		}
+
+		name := d.Get("name").(string)
+		status := d.Get("status").([]interface{})
+        state := status[0].(map[string]interface{})["state"]
+        log.Printf("[DEBUG] Connection %q has state %v.", name, state)
+	    if state == "CREATING" {
+			return resource.RetryableError(fmt.Errorf("Connection %q has state %q.", name, state))
+		}
+		return nil
```

Comment:
Currently, if state == ERROR we will return nil but not error which is the same as considering it as success. Likely, we want something similar like
```suggestion
		if state == "CREATING" || state == "UPDATING" { // add states that indicates the action is ongoing
			return resource.RetryableError(fmt.Errorf("Connection %q has state %q.", name, state))
		} else if state == "ERROR" || state == "AUTHORIZATION_REQUIRED" { // add states that indicates the action fails
			return resource.NonRetryableError(fmt.Errorf("Connection %q has state %q.", name, state))
		}
		log.Printf("[DEBUG] Connection %q has state %q.", name, state)
		return nil
```

Reply:
We wouldn't want to call the resource creation or updation is in error state, just that the connection to backend is in error state, we wouldn't want to block that because of it, we are okay with the connection state is in error state, we would consider the creation or updation call to be successful and this check is to tell terraform that only after this, consider the operation to be done.

Reply:
When the creation of resource fails, the LRO will return an error, which is handled already. The state error doesn't mean the creation of resource failed.

Reply:
Can customers use the resource if they're in error state? What are they expect to do with these resources?

Reply:
As a product, we enable users to connect to their backends and trigger operations on their backends. So this state, tells us about the state of connection to that backend and the state of resource itself.


The state reconciles every 10 seconds, the credentials can be valid or the customer's backend could be up later, so they can use it.

Reply:
The usage of these resources are by another product called "Application integration"

Reply:
Okay, so if users provide invalid configs or creation/update run into issues, LRO will return errors. So status.state does not indicate creation/update fails, it's just LRO sometimes returns early when the resource is still creating/updating, we need to wait until it's done?

Reply:
Yes. Exactly.